src/StackExchange.Redis/RedisSubscriber.cs

```diff
@@ -35,273 +55,178 @@ internal bool GetSubscriberCounts(in RedisChannel channel, out int handlers, out
         {
-            Task last = null;
-            lock (subscriptions)
+            foreach (var pair in subscriptions)
             {
-                foreach (var pair in subscriptions)
-                {
-                    pair.Value.MarkCompleted();
-                    var task = pair.Value.UnsubscribeFromServer(pair.Key, flags, asyncState, false);
-                    if (task != null) last = task;
-                }
-                subscriptions.Clear();
+                pair.Value.UpdateServer();
             }
-            return last ?? CompletedTask<bool>.Default(asyncState);
         }
 
-        internal Task RemoveSubscription(in RedisChannel channel, Action<RedisChannel, RedisValue> handler, ChannelMessageQueue queue, CommandFlags flags, object asyncState)
+        /// <summary>
+        /// Ensures all subscriptions are connected to a server, if possible.
+        /// </summary>
+        internal void EnsureSubscriptions(CommandFlags flags = CommandFlags.None)
         {
-            Task task = null;
-            lock (subscriptions)
+            // TODO: Subscribe with variadic commands to reduce round trips
+            foreach (var pair in subscriptions)
             {
-                if (subscriptions.TryGetValue(channel, out Subscription sub))
-                {
-                    bool remove;
-                    if (handler == null & queue == null) // blanket wipe
-                    {
-                        sub.MarkCompleted();
-                        remove = true;
-                    }
-                    else
-                    {
-                        remove = sub.Remove(handler, queue);
-                    }
-                    if (remove)
-                    {
-                        subscriptions.Remove(channel);
-                        task = sub.UnsubscribeFromServer(channel, flags, asyncState, false);
-                    }
-                }
+                DefaultSubscriber.EnsureSubscribedToServer(pair.Value, pair.Key, flags, true);
             }
-            return task ?? CompletedTask<bool>.Default(asyncState);
         }
 
-        internal void ResendSubscriptions(ServerEndPoint server)
+        /// <summary>
+        /// Ensures all subscriptions are connected to a server, if possible.
+        /// </summary>
+        internal async Task<long> EnsureSubscriptionsAsync(CommandFlags flags = CommandFlags.None)
         {
-            if (server == null) return;
-            lock (subscriptions)
+            // TODO: Evaluate performance here, this isn't good for a large number of subscriptions.
+            // It's probable we want to fire and forget `n` here, recording how many are going to try to reconnect?
+            long count = 0;
+            foreach (var pair in subscriptions)
             {
-                foreach (var pair in subscriptions)
+                if (await DefaultSubscriber.EnsureSubscribedToServerAsync(pair.Value, pair.Key, flags, true))
```

Comment:
This is currently *not* called as `FireAndForget` upstream in the connection pipe and is a `O(n)` issue. We can do `FireAndForget` all the way down, but there's a subtle issue of a default task state for `<bool>` being false and the count would always be zero - would need to change implementation somehow to get a real count out of it for the connection log trace.